snuba/utils/streams/streaming.py

```diff
@@ -86,6 +107,292 @@ def join(self, timeout: Optional[float] = None) -> None:
         self.__next_step.join(timeout)
 
 
+# A serialized message is composed of a pickled ``Message`` instance (bytes)
+# and a sequence of ``(offset, length)`` that referenced locations in a shared
+# memory block for out of band buffer transfer.
+SerializedMessage = Tuple[bytes, Sequence[Tuple[int, int]]]
+
+
+class ValueTooLarge(ValueError):
+    """
+    Raised when a value is too large to be written to a shared memory block.
+    """
+
+
+class MessageBatch(Generic[TPayload]):
+    def __init__(self, block: SharedMemory) -> None:
+        self.block = block
+        self.__items: MutableSequence[SerializedMessage] = []
+        self.__offset = 0
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {len(self)} items, {self.__offset} bytes>"
+
+    def __len__(self) -> int:
+        return len(self.__items)
+
+    def __getitem__(self, index: int) -> Message[TPayload]:
+        data, buffers = self.__items[index]
+        return pickle.loads(
+            data,
+            buffers=[
+                self.block.buf[offset : offset + length].tobytes()
```

Comment:
Yeah I agree who uses this class does not care about it. I am thinking to how we could preserve the invariants around what gets copied and what does not in this whole design when people will try to refactor or make changes. Adding a copy of the message by mistake could be hard to spot when testing locally, and removing a copy by mistake could be hard to spot if none is modifying the underlying value.

In this case actually I think the easiest way, is to just have a an assertion in test_message_batch (line 146) that checks that the payloads are not the same object. So we are stating that it is intentional (which is not the general behavior of most collections with a `__getitem__` method) and none will think of improving the performance by removing that copy.

Reply:
I can add the test if you feel strongly about this, but I'm not sure the invariant (or lack thereof) really matters, because the successful execution of the program (at least as it stands today) doesn't depend on these buffers being either a copy or not a copy and there aren't any guarantees being made here about one approach or the other.

I'm not sure I follow the point about `__getitem__` since that would only apply to the `Message[TPayload]` object, which is always going to be a copy, since it is read from the pickle stream (unless we change that behavior in the future, at a distance from `MessageBatch`, to allow the _entire_ `Message` instance to be able to be entirely serialized into a byte buffer.)